### PR TITLE
Write wiki-like documentation pages

### PR DESCRIPTION
**Pull request by gselzer**
This PR introduces a set of drafts for wiki-like documentation for using the SciJava Ops/ImageJ Ops2 framework, addressing the points of need described in scijava/scijava#124.

@hinerm let me know if there's anything you think I left out. We also have to figure out where they should live.

### CODE REVIEW

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/CallingOps.md]: Can you add an example on what chaining `airtyX()` would look like? I'm guessing something like:

```java
ops.op("filter.gauss").arityX().arityX().arityX()
```
Can you also mix and match like this?

```java
ops.op("filter.gauss").arity2().arityX()
```

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/CallingOps.md]: Typo? Should read `gaussOp.compute(inImage)`?

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:37]: I find migration documents that have numbered lists easier to follow. From the reader's perspective I want to see what steps I need to do to get my Op converted...and I'm probably irritated! There are only two steps here, but having the numbers shows what few steps you need and calls attention to the last line about remove the dependency on ImageJ Ops.

I'd recommend adding a numbered list here.

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:1]: Overall great document. I think it could benefit from the headings being numbered.

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Use another word for the second "Therefore" or replace with "Thus".

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Remove extra space?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Remove extra space.

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Consider section name "Neighborhood-wise Ops" to match the previous section.

File: imagej/imagej-ops2-tutorial/docs/SearchingForOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/SearchingForOps.md:1]: Is there a way to generate a tree or list output of all ops that are available to the environment? If so, that would be a nice section to add to this doc.

File: imagej/imagej-ops2-tutorial/docs/SearchingForOps.md
Comment (elevans) [inline imagej/imagej-ops2-tutorial/docs/SearchingForOps.md]: What does the output look like from `ops.descriptions("filter.gauss")`? Can you add that here?

**Review by elevans — COMMENTED** [review]
Good stuff! I learned a few things about the op builder which I found very useful.

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: I hesitate to answer that directly, because I'm hoping it's obvious from the wiki page :sweat_smile: 

Please tell me your understanding after reading the change made in bb7aa138, @elevans!

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: Yes, good catch! Fixed with d83281d8

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:37]: Converted to a numbered list in 16554321

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:1]: Let me know what you think of 16554321

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: I chose to just remove it in 60d278fb

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Removed in 8c5caa33

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Removed in 8c5caa33

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Changed in ffd356ae

File: imagej/imagej-ops2-tutorial/docs/SearchingForOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/SearchingForOps.md]: Added in the descriptions in 46502f0d

File: imagej/imagej-ops2-tutorial/docs/SearchingForOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/SearchingForOps.md:1]: You could use `ops.descriptions()` (no-args ctor) - I *really* don't want to include that printout here, though - it'd be way too long.

I did add a note in e8c84471

**Comment by gselzer** [comment]
Thanks for the review @elevans - I think I addressed all of your comments. Let me know if you see anything else!

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (elevans) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: This is great and I see my misunderstanding. Can you add "...we do this by adding on of the `arityX()` methods, **where `X` is the number of inputs**, to chain...".

I know this is pretty minor but the `code` highlighting of `arityX()` actually made me think there is an `arityX()` that you can use!

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (elevans) [reply imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:1]: Love it! This feels like a nice step-by-step migration document.

**Comment by elevans** [comment]
I think this is great! Ready to merge :+1:

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (ctrueden) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: Maybe `arity*` would be clearer, since `*` is not a valid character for method names.

Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: > Maybe `arity*` would be clearer, since `*` is not a valid character for method names.

That's a great idea - I did it with 0e461ce8

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/CallingOps.md]: This does not compile for me, and it's not clear to me how it could even work.. shouldn't it be `gaussOp.compute(inImage, 2.0, outImage)`?

I think this section should be a modified to incrementally build up modifications to the original call:
1. the option to use `unary`, `binary`, etc.. in place of `ops.op(..).arity*()`
2. the option to use `inType` and `outType`
3. Using `Nil` with `inType` and `outType`
4. Using `compute()` and Op reuse
5. Model the `Function` and `Inplace` alternatives

Verify the code is working as expected in an IDE.

There should also be a `warnings` or `pitfalls` or something section. e.g. I pasted this into `OpsIntro` and it didn't compile when using `img` as the `inImage` because it's an `Img<?>`, and using a wildcard in the `input` method is a recipe for disaster I think.

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:35]: Does this all need to change after https://github.com/scijava/incubator/pull/79?

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:5]: A "Migrating from ImageJ Ops" section is great. Presumably this step is a requisite for any implementer of ops. So if we're going to document migration we might as well have a page about implementation - then this pom-scijava info can be migrated there and reference from the migration doc.

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: lots of newlines?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: "iterating.op"?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: "neighborhood.op"

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: name should be `my.op2` but I think let's just change the names of all the ops in this demo to something descriptive and easily differentiated

"collection.op" or "double.op"?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: PixelwiseOp is used above... `MyCollectionOp`?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: elementwise Op, perhaps?

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: "pixel.op"

File: imagej/imagej-ops2-tutorial/docs/ScriptingInImageJ2.md
Comment (hinerm) [inline imagej/imagej-ops2-tutorial/docs/ScriptingInImageJ2.md:48]: is `new Double` actually required?

**Review by hinerm — CHANGES_REQUESTED** [review]
@gselzer Nice work. I especially like the migration walkthrough. I'd just like to see a couple sections expanded and revised.

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Removed in 640400e1

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Good idea - renamed in f4f121e2

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Renamed in f4f121e2

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Renamed in f4f121e2

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Renamed in f4f121e2 to `MyElementOp` - I didn't want any confusion with `OpCollection`s.

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Renamed in f4f121e2

File: imagej/imagej-ops2-tutorial/docs/OpCurrency.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/OpCurrency.md]: Renamed in f4f121e2

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:35]: Yeah, it will have to change, in one of two ways:
1. We move this block upstream, like we did for [therapi](https://github.com/scijava/pom-scijava-base/pull/31) - in this case, we'll probably need to add some opt-in flag to POMs.
2. We don't move this block upstream, and we simply change the annotation processor to use the new `scijava-javadoc-parser` module, like we do in that PR.

Either way, we should either (a) not merge this until #79 merges, or (b) file an issue to change it later.

File: imagej/imagej-ops2-tutorial/docs/CallingOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/CallingOps.md]: > This does not compile for me, and it's not clear to me how it could even work.. shouldn't it be `gaussOp.compute(inImage, 2.0, outImage)`?

Yup, good catch!

> 1. the option to use `unary`, `binary`, etc.. in place of `ops.op(..).arity*()`

Sure, I added that section in 86b6132

> 2. the option to use `inType` and `outType`

Man, I'm really hesitant to put that in what should be the most approachable page of the docs - using `Type`s for matching is *so* verbose and complex.

...honestly, what do we *really* use `inType`/`outType` patterns *for* nowadays? The only thing I think *might* be useful is `outType` for function matching - we don't use `inType`/`outType` in the ImageJ Ops2 tests *at all*.

@hinerm @ctrueden I think it would be really worthwhile to have a discussion about removing the `inType`/`outType` syntax from the OpBuilder (of course the `OpEnvironment` still needs the Type/`Nil` syntax) - we could potentially remove a lot the `OpBuilder` autocomplete clutter

Needless to say, I haven't added anything past 2 either.

> Verify the code is working as expected in an IDE.

The existing code compiles now

> There should also be a warnings or pitfalls or something section. e.g. I pasted this into OpsIntro and it didn't compile when using img as the inImage because it's an Img<?>, and using a wildcard in the input method is a recipe for disaster I think.

Good idea - added a `Common Pitfalls` section.

File: imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/MigratingFromImageJOps.md:5]: @hinerm I wrote a start in 2ea41a72.

I'd like to write an example project that we can direct people to, although that may come along with the work that we're doing in other PRs. I'd like to get those merged (#73, #79, etc) first.

File: imagej/imagej-ops2-tutorial/docs/ScriptingInImageJ2.md
Comment (gselzer) [reply imagej/imagej-ops2-tutorial/docs/ScriptingInImageJ2.md:48]: Unfortunately, it seems to be - otherwise, they become `BigDecimal`s

**Comment by gselzer** [comment]
@hinerm thanks for the review! I think I've addressed all of your comments - let me know what you think!

**Review by hinerm — APPROVED** [review]

**Pull request merged**
